drivers: regulator: print regulator tree summary

etienne-lms opened the pull request:
Replaces existing `regulator_print_state()` with `regulator_print_summary()` on the model of what's been done for clocks (https://github.com/OP-TEE/optee_os/pull/6474).
The function is renamed for consistency.

etienne-lms commented:
Rebased

etienne-lms commented:
I've appended a commit for adding a command in the statistics PTA to allow non-secure to request print of clock & regulator trees on the secure console: see commit _"pta: stats: add commands to print clock and regulator trees"_.

CI / make (multi-platform) test failure should be fixed with https://github.com/OP-TEE/optee_os/pull/6484.

jforissier commented on core/pta/stats.c:
```diff
@@ -51,6 +53,16 @@
+ *
+ * [in]    value[0].a        Target driver, one of STATS_DRIVER_TYPE_*
+ */
+#define STATS_CMD_PRINT_DRIVERS_INFO	5
```
`_DRIVER_` (singular)? since the command prints for one driver

etienne-lms replied:
ok, makes sense. I'll address that.

jforissier commented on core/pta/stats.c:
```diff
@@ -209,6 +221,28 @@ static TEE_Result get_system_time(uint32_t type,
 	return TEE_SUCCESS;
 }
 
+static TEE_Result get_drivers_info(uint32_t type, TEE_Param p[TEE_NUM_PARAMS])
```
`print_driver_info()`. This does not get anything.
...which leads to the question: should a buffer be returned to the client instead? That would be more consistent with other statistics and also would be usable when `CFG_TEE_CORE_LOG_LEVEL < TRACE_INFO`.

etienne-lms replied:
Right, I'll change the function label.
Note that command `STATS_CMD_MEMLEAK_STATS` already behaves this way.

jforissier replied:
> Note that command `STATS_CMD_MEMLEAK_STATS` already behaves this way.

Fair enough. Printing to the secure console is probably good enough for the main use cases (debugging).

etienne-lms commented:
I've addressed the several CI make error and address another potential build error: cases when `CFG_DRIVERS_{CLK|REGUALTOR}` is disabled..
I've rebased on master branch tip to leverage https://github.com/OP-TEE/optee_os/pull/6484 fix.

jenswi-linaro commented on core/drivers/regulator/regulator.c:
```diff
@@ -323,9 +303,116 @@ static TEE_Result regulator_core_cleanup(void)
+	SLIST_FOREACH(regulator, &regulator_device_list, link) {
+		if (regulator->supply == root) {
+			print_regulator(regulator, indent + 1);
+			print_regulator_subtree(regulator, indent + 1);
```
Would it be possible to do this without recursion?

jenswi-linaro replied:
Yes, that's why recursion is discouraged. A comment explaining why this will not be a problem here might be sufficient. Asserting that `indent` isn't too large could also be an option.

etienne-lms replied:
I'm ok with setting an upper limit to `indent`. Implementing this without recursion looks a bit painfull.
I'll append a fixup commit and add the same to `clk_print_tree()`.

jenswi-linaro replied:
> Implementing this without recursion looks a bit painfull.

Oh, that sounds almost like a challenge. ;-)
You have parent pointers in the nodes so there's no need to save a list of pointers to find the path back to the root.

etienne-lms replied:
It would be a bit of code to create the ascending tree of devices. Do you think it's worth it? This recursion is less than 10 lines, it like it. That said, I really understand your concern.

etienne-lms replied:
I tried but I always end up with some sort of recursive calls to descend the tree to print the clock info.
The sole way out would be to allocate some big reference matrix to track per item / per level state of what has been printed so far. I found that quite ugly.

I would be in favor for the recursive calls, bounded to 16 levels (32 seems to high from the trace found on clock: https://github.com/OP-TEE/optee_os/pull/6489#issuecomment-1824730803).

jenswi-linaro replied:
This should do the trick:
``` C
static struct regulator *find_next_regulator(struct regulator *parent,
					     struct regulator *sibling)
{
	struct regulator *r = NULL;

	if (sibling)
		r = SLIST_NEXT(sibling, link);
	else
		r = SLIST_FIRST(&regulator_device_list);

	while (r && r->supply != parent)
		r = SLIST_NEXT(r, link);

	return r;
}

static void regulator_print_tree(void)
{
	struct regulator *regulator = NULL;
	struct regulator *parent = NULL;
	struct regulator *next = NULL;
	int indent = -1;

	while (true) {
		next = find_next_regulator(parent, regulator);
		if (next) {
			print_regulator(next, indent + 1);
			/* Enter the subtree of the next regulator */
			parent = next;
			indent++;
			regulator = NULL;
		} else {
			/*
			 * We've processed all children at this level.
			 * If parent is NULL we're at the top and are done.
			 */
			if (!parent)
				break;
			/*
			 * Move up one level to resume with the next
			 * regulator of the parent.
			 */
			regulator = parent;
			parent = parent->supply;
			indent--;
		}
	}
}
```

jenswi-linaro commented on core/drivers/regulator/regulator.c:
```diff
@@ -323,9 +303,116 @@ static TEE_Result regulator_core_cleanup(void)
+	IMSG("%s", msg_buf);
+}
+
+static void print_regulator_subtree(struct regulator *root __maybe_unused,
```
Wouldn't `parent` be a more suitable name than `root`?

etienne-lms replied:
true. I'll change that.

etienne-lms commented:
Regarding CI make test failure, i've created an Issue entry to get some help.

etienne-lms commented:
Comments addressed.

I've created a specific P-r (https://github.com/OP-TEE/optee_os/pull/6481) for the variable sized field issue reported by Clang. I picked the commit in this series to have CI test running (hopefully) without this issue.

Regarding the recursive call to print the regulator tree, I've addressed the same concern in existing `clk_print_tree()` with P-R https://github.com/OP-TEE/optee_os/pull/6489.

etienne-lms commented:
I think i've put too many things in this P-R (regulator tree, regulator fixup, clock tree fixup, print trees from PTA stats).
I'll leave this P-R as is for the moment and create dedocated P-Rs per topic:
- regulator tree: see https://github.com/OP-TEE/optee_os/pull/6498
- regulator fixup regarding variable sized field: see https://github.com/OP-TEE/optee_os/pull/6487
- clock tree fixes: see https://github.com/OP-TEE/optee_os/pull/6497
- print trees from a statistics PTA command: TODO
